python/pyarrow/tests/test_dataset.py

```diff
@@ -0,0 +1,364 @@
+            ds.CompareOperator.Equal,
+            ds.FieldExpression('part'),
+            ds.ScalarExpression(1)
+        ),
+        'subdir/2/xxx/file1.parquet': ds.ComparisonExpression(
+            ds.CompareOperator.Equal,
+            ds.FieldExpression('part'),
+            ds.ScalarExpression(2)
+        )
+    }
+    source = ds.FileSystemDataSource(
+        mockfs,
+        paths,
+        source_partition=source_partition,
+        path_partitions=path_partitions,
+        file_format=file_format
+    )
+    assert source.partition_expression.equals(source_partition)
+
+
+def test_dataset(simple_data_source, tree_data_source, record_batch, schema):
+    dataset = ds.Dataset([simple_data_source, tree_data_source], schema)
+
+    assert isinstance(dataset, ds.Dataset)
+    assert isinstance(dataset.schema, pa.Schema)
+    assert isinstance(dataset.sources[0], ds.SimpleDataSource)
+    assert isinstance(dataset.sources[1], ds.TreeDataSource)
+
+    condition = ds.ComparisonExpression(
+        ds.CompareOperator.Equal,
+        ds.FieldExpression('i64'),
+        ds.ScalarExpression(1)
+    )
+    # TODO(kszucs): test non-boolean expressions for filter do raise
+    builder = dataset.new_scan()
+    assert isinstance(builder, ds.ScannerBuilder)
+    assert builder.schema.equals(schema)
+
+    scanner = builder.finish()
+    assert isinstance(scanner, ds.Scanner)
+    assert len(list(scanner.scan())) == 3
+
+    for task in scanner.scan():
+        assert isinstance(task, ds.ScanTask)
+        for batch in task.scan():
+            assert batch.equals(record_batch)
+
+    table = scanner.to_table()
+    assert isinstance(table, pa.Table)
+    assert len(table) == 15
+
+    scanner = dataset.new_scan().use_threads(True).filter(condition).finish()
+    result = scanner.to_table()
+    expected = pa.table([[1] * 3, [1.0] * 3], schema=schema)
+    assert result.equals(expected)
+
+
+def test_scanner_builder(dataset):
+    context = ds.ScanContext()
+    builder = ds.ScannerBuilder(dataset, context)
+    scanner = builder.finish()
+    assert isinstance(scanner, ds.Scanner)
+    scanner.scan()
+
+    builder = dataset.new_scan()
+    builder.project(['i64'])
+    scanner = builder.finish()
+    scanner.scan()
+    assert isinstance(scanner, ds.Scanner)
+
+
+def test_abstract_classes():
+    classes = [
+        ds.FileFormat,
+        ds.Scanner,
+        ds.DataSource,
+        ds.Expression,
+        ds.PartitionScheme,
+    ]
+    for klass in classes:
+        with pytest.raises(TypeError):
+            klass()
+
+
+def test_partition_scheme(schema):
+    for klass in [ds.SchemaPartitionScheme, ds.HivePartitionScheme]:
+        scheme = klass(schema)
+        assert isinstance(scheme, ds.PartitionScheme)
+
+    scheme = ds.SchemaPartitionScheme(
+        pa.schema([
+            pa.field('group', pa.int64()),
+            pa.field('key', pa.float64())
+        ])
+    )
+    expr = scheme.parse('/3/3.14')
```

Comment:
The initial `/` seemed weird, is that required?

Reply:
Nope